engine/common/provider/engine.go

```diff
@@ -73,81 +123,57 @@ func New(log zerolog.Logger, metrics module.EngineMetrics, net network.Network,
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
-}
+	cm := component.NewComponentManagerBuilder()
+	cm.AddWorker(e.processQueuedRequestsShovellerWorker)
+	for i := uint(0); i < requestWorkers; i++ {
+		cm.AddWorker(e.processEntityRequestWorker)
+	}
 
-// Submit submits the given message from the node with the given origin ID
-// for processing in a non-blocking manner. It returns instantly and logs
-// a potential processing error internally when done.
-func (e *Engine) Submit(channel channels.Channel, originID flow.Identifier, message interface{}) {
-	e.unit.Launch(func() {
-		err := e.Process(channel, originID, message)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
-}
+	e.cm = cm.Build()
+	e.Component = e.cm
 
-// ProcessLocal processes an message originating on the local node.
-func (e *Engine) ProcessLocal(message interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(e.me.NodeID(), message)
-	})
+	return e, nil
 }
 
 // Process processes the given message from the node with the given origin ID in
 // a blocking manner. It returns the potential processing error when done.
-func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, message interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, message)
-	})
-}
-
-// process processes events for the propagation engine on the consensus node.
-func (e *Engine) process(originID flow.Identifier, message interface{}) error {
+func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, event interface{}) error {
+	select {
+	case <-e.cm.ShutdownSignal():
+		e.log.Warn().
+			Hex("origin_id", logging.ID(originID)).
+			Msgf("received message after shutdown")
+		return nil
+	default:
+	}
 
 	e.metrics.MessageReceived(e.channel.String(), metrics.MessageEntityRequest)
-	defer e.metrics.MessageHandled(e.channel.String(), metrics.MessageEntityRequest)
 
-	e.unit.Lock()
-	defer e.unit.Unlock()
-
-	switch msg := message.(type) {
-	case *messages.EntityRequest:
-		return e.onEntityRequest(originID, msg)
-	default:
-		return engine.NewInvalidInputErrorf("invalid message type (%T)", message)
+	err := e.requestHandler.Process(originID, event)
+	if err != nil {
+		if engine.IsIncompatibleInputTypeError(err) {
+			e.log.Warn().
+				Hex("origin_id", logging.ID(originID)).
+				Str("channel", channel.String()).
+				Str("event", fmt.Sprintf("%+v", event)).
+				Msgf("received unsupported message type")
+			return nil
+		}
+		return fmt.Errorf("unexpected error while processing engine event: %w", err)
 	}
+
+	return nil
 }
 
-func (e *Engine) onEntityRequest(originID flow.Identifier, req *messages.EntityRequest) error {
-	lg := e.log.With().Uint64("nonce", req.Nonce).Str("origin_id", originID.String()).Logger()
+// onEntityRequest processes an entity request message from a remote node.
+// All errors returned by this function are benign and should not cause the engine to crash.
```

Comment:
This kind of error handling is OK, only as an exception from the general rule, for example for side-effect-free functions. Here, and as a general rule, it is dangerous, because it can hide unexpected failure conditions. 

For example:
* any error when requesting identities for the latest finalized block, is not expected
* an error besides NotFound is not expected for the `retrieve` function

It's challenging to update our existing code when dependencies don't have error returns documented. Here's how I suggest we approach it:

**Optimal:** When updating / creating a new function, update the error documentation of the sub-functions called by that function, then implement error handling for documented errors. Then we can correctly document errors for the new/updated function, and don't leave technical debt for later.

If it's too onerous to update error documentation for every sub-function, then for sub-functions we don't update:
* Add a TODO to the sub-function godoc noting that the function's errors must be documented
* If possible, implement error handling at the call-site of the sub-function, so that the new/updated function can have exhaustive error documentation, despite the sub-function, for example by:
  * wrapping errors from the sub-function
  * document reasoning for why certain errors are expected or not

**Backup:** If we aren't going to fix error handling for an updated function, we should add a TODO to the updated function noting the tech debt that errors must be documented. 

```suggestion
// Error returns:
// * InvalidInputError if the list of requested entities is invalid (empty)
// * generic error in case of unexpected failure or implementation bug
```